Intersect.Client/Interface/Game/EventWindow.cs

```diff
@@ -317,6 +384,140 @@ public void EventResponse1_Clicked(Base sender, ClickedEventArgs arguments)
             base.Hide();
         }
 
+        private void SkipTypewriting()
+        {
+            if (Writer?.Done ?? true)
+            {
+                return;
+            }
+
+            Writer.End();
+        }
     }
 
+    sealed class Typewriter
+    {
+        readonly long TypingSpeed = ClientConfiguration.Instance.TypewriterLetterSpeed;
+        readonly long FullStopSpeed = ClientConfiguration.Instance.TypewriterFullstopSpeed;
+        readonly long PartialStopSpeed = ClientConfiguration.Instance.TypewriterPartialstopSpeed;
+
+        private List<Label> Labels { get; set; }
+        private string[] Lines { get; set; }
+        private Label CurrentLabel => Labels.ElementAtOrDefault(LineIdx);
+        private string CurrentLine => Lines.ElementAtOrDefault(LineIdx);
+        private int LineIdx { get; set; }
+        private int CharIdx { get; set; }
+        private char? LastChar { get; set; }
+        private long LastUpdateTime;
+        private long LastPauseTime { get; set; }
+
+        public bool Done { get; private set; }
+
+        public long DoneAt { get; set; }
+
+        public void CarriageReturn()
+        {
+            LineIdx++;
+            if (LineIdx >= Lines.Length)
+            {
+                End();
+                return;
+            }
+            CharIdx = 0;
+            LastChar = null;
+        }
+
+        public void Initialize(List<Label> labels)
+        {
+            LastUpdateTime = Timing.Global.Milliseconds;
+            Labels = labels;
+            Lines = Labels.Select(l => l.Text).ToArray();
+            Labels.ForEach(l => l.SetText("")); // Clear the text from the labels. The writer is the captain now
+            LineIdx = 0;
+            CharIdx = 0;
+            LastChar = null;
+            Done = false;
+        }
+
+        public void Write(string voice)
+        {
+            if (Done)
+            {
+                return;
+            }
+
+            if (string.IsNullOrEmpty(CurrentLine) || CurrentLabel == default)
```

Comment:
Is there a reason you are doing IsNullOrEmpty instead of IsNullOrWhiteSpace?

Reply:
Isn't nullOrEmpty more inclusive? Changed regardless

Reply:
No, == null only checks for null, NullOrEmpty checks for null/`""`, NullOrWhitespace checks for that, plus all whitespace.